File: server/scripts/postfix-pipe.py
import sys
import os
import json
import time
from email import policy
from email.parser import BytesParser

DB_PATH = '/var/www/xmorf/server/data/db.json'
UPLOAD_DIR = '/var/www/xmorf/server/uploads'

os.makedirs(UPLOAD_DIR, exist_ok=True)

def format_size(size_bytes):
    if size_bytes < 1024:
        return f"{size_bytes} B"
    elif size_bytes < 1024 * 1024:
        return f"{size_bytes / 1024:.1f} KB"
    else:
        return f"{size_bytes / (1024 * 1024):.1f} MB"

def get_file_type(filename):
    ext = filename.split('.')[-1].lower() if '.' in filename else ''
    if ext in ['png', 'jpg', 'jpeg', 'gif', 'webp', 'svg']:
        return 'image'
    elif ext == 'pdf':
        return 'pdf'
    elif ext in ['zip', 'rar', '7z', 'tar', 'gz']:
        return 'archive'
    return 'document'

def main():
    try:
        raw_bytes = sys.stdin.buffer.read()
        if not raw_bytes:
            return

        msg = BytesParser(policy=policy.default).parsebytes(raw_bytes)

        sender_header = str(msg.get('From', 'external@xmorf.net'))
        sender_email = sender_header
        sender_name = 'External Sender'

        if '<' in sender_header and '>' in sender_header:
            sender_name = sender_header.split('<')[0].strip(' "\'')
            sender_email = sender_header.split('<')[1].split('>')[0].strip()
        else:
            sender_email = sender_header.strip(' "\'')
            sender_name = sender_email.split('@')[0] if '@' in sender_email else sender_email

        recipient_arg = sys.argv[1] if len(sys.argv) > 1 else str(msg.get('To', 'demo@xmorf.net'))
        recipient = recipient_arg.replace('<', '').replace('>', '').strip().lower()

        subject = str(msg.get('Subject', '(No Subject)'))

        body = ""
        body_part = msg.get_body(preferencelist=('plain', 'html'))
        if body_part:
            body = body_part.get_content()
        else:
            body = "(No readable content)"

        attachments = []
        for part in msg.walk():
            if part.get_content_maintype() == 'multipart':
                continue
            filename = part.get_filename()
            if filename:
                unique_fn = f"{int(time.time()*1000)}-{os.urandom(4).hex()}-{filename}"
                file_path = os.path.join(UPLOAD_DIR, unique_fn)
                payload = part.get_payload(decode=True)
                if payload:
                    with open(file_path, 'wb') as f:
                        f.write(payload)
                    attachments.append({
                        "id": f"att-{int(time.time()*1000)}-{os.urandom(2).hex()}",
                        "name": filename,
                        "size": format_size(len(payload)),
                        "type": get_file_type(filename),
                        "filename": unique_fn,
                        "url": f"/api/uploads/{unique_fn}"
                    })

        # Load db.json
        db = {"users": [], "emails": []}
        if os.path.exists(DB_PATH):
            try:
                with open(DB_PATH, 'r', encoding='utf-8') as f:
                    db = json.load(f)
            except Exception:
                pass

        if 'emails' not in db or not isinstance(db['emails'], list):
            db['emails'] = []

        import datetime
        now_str = datetime.datetime.now().strftime("%H:%M")

        new_email = {
            "id": f"em-inbound-{int(time.time()*1000)}-{os.urandom(2).hex()}",
            "ownerEmail": recipient,
            "folder": "inbox",
            "senderName": sender_name or sender_email,
            "senderEmail": sender_email.lower(),
            "recipient": recipient,
            "subject": subject,
            "preview": (body[:90].replace('\r', ' ').replace('\n', ' ') + '...') if body else 'No content',
            "body": body,
            "timestamp": "Just now",
            "date": now_str,
            "isUnread": True,
            "isStarred": False,
            "attachments": attachments
        }

        db['emails'].insert(0, new_email)

        with open(DB_PATH, 'w', encoding='utf-8') as f:
            json.dump(db, f, indent=2, ensure_ascii=False)

        print(f"SUCCESS: Saved email for {recipient} from {sender_email}")
    except Exception as e:
        print(f"ERROR: {e}", file=sys.stderr)

if __name__ == '__main__':
    main()

File: server/scripts/postfix-pipe.js
const fs = require('fs');
const path = require('path');
const { simpleParser } = require('mailparser');

const DB_PATH = path.join(__dirname, '../data/db.json');
const UPLOAD_DIR = path.join(__dirname, '../uploads');

if (!fs.existsSync(UPLOAD_DIR)) {
  fs.mkdirSync(UPLOAD_DIR, { recursive: true });
}

let rawEmail = '';
process.stdin.setEncoding('utf8');
process.stdin.on('data', chunk => { rawEmail += chunk; });
process.stdin.on('end', async () => {
  try {
    const parsed = await simpleParser(rawEmail);

    const senderEmail = (parsed.from && parsed.from.value && parsed.from.value[0] && parsed.from.value[0].address)
      ? parsed.from.value[0].address.toLowerCase().trim()
      : 'external@xmorf.net';

    const senderName = (parsed.from && parsed.from.value && parsed.from.value[0] && parsed.from.value[0].name)
      ? parsed.from.value[0].name
      : senderEmail.split('@')[0];

    let recipient = process.argv[2] || (parsed.to && parsed.to.value && parsed.to.value[0] ? parsed.to.value[0].address : 'demo@xmorf.net');
    recipient = recipient.replace(/^<|>$/g, '').toLowerCase().trim();

    const subject = parsed.subject || '(No Subject)';
    const body = parsed.text || parsed.html || 'No content';

    const attachments = [];
    if (parsed.attachments && parsed.attachments.length > 0) {
      for (const att of parsed.attachments) {
        const uniqueFilename = Date.now() + '-' + Math.round(Math.random() * 1e9) + '-' + (att.filename || 'file');
        const filePath = path.join(UPLOAD_DIR, uniqueFilename);
        fs.writeFileSync(filePath, att.content);

        const formatSize = (bytes) => {
          if (bytes < 1024) return bytes + ' B';
          if (bytes < 1024 * 1024) return (bytes / 1024).toFixed(1) + ' KB';
          return (bytes / (1024 * 1024)).toFixed(1) + ' MB';
        };

        const fileExt = path.extname(att.filename || '').toLowerCase().replace('.', '');
        let fileType = 'document';
        if (['png', 'jpg', 'jpeg', 'gif', 'webp', 'svg'].includes(fileExt)) fileType = 'image';
        if (fileExt === 'pdf') fileType = 'pdf';
        if (['zip', 'rar', '7z', 'tar', 'gz'].includes(fileExt)) fileType = 'archive';

        attachments.push({
          id: 'att-' + Date.now() + '-' + Math.round(Math.random() * 1000),
          name: att.filename || 'attachment',
          size: formatSize(att.size || att.content.length),
          type: fileType,
          filename: uniqueFilename,
          url: `/api/uploads/${uniqueFilename}`
        });
      }
    }

    // Insert into db.json
    let db = { users: [], emails: [] };
    if (fs.existsSync(DB_PATH)) {
      try {
        db = JSON.parse(fs.readFileSync(DB_PATH, 'utf8'));
      } catch(e){}
    }
    if (!Array.isArray(db.emails)) db.emails = [];

    const newInboxEmail = {
      id: 'em-inbound-' + Date.now() + '-' + Math.floor(Math.random() * 1000),
      ownerEmail: recipient,
      folder: 'inbox',
      senderName: senderName,
      senderEmail: senderEmail,
      recipient: recipient,
      subject: subject,
      preview: body ? body.substring(0, 90).replace(/[\r\n]+/g, ' ') + '...' : 'No content',
      body: body,
      timestamp: 'Just now',
      date: new Date().toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' }),
      isUnread: true,
      isStarred: false,
      attachments: attachments
    };

    db.emails.unshift(newInboxEmail);
    fs.writeFileSync(DB_PATH, JSON.stringify(db, null, 2), 'utf8');
    console.log(`Inbound email successfully saved to inbox for ${recipient} from ${senderEmail}`);
  } catch (err) {
    console.error('Error processing inbound email pipe:', err);
  }
});
